ESP32-S2: Basic MCUboot support

almir-okato opened the pull request:
This PR adds basic support for ESP32-S2 SoC from Espressif Systems.

**Notes:** 
- Both Zephyr and NuttX RTOSes for ESP32-S2 would be able to boot in the future, however the changes in their respective repositories still are WIP and aren't upstream yet. 

## Setting up the environment
- `git submodule update --init --recursive` (Update ESP-IDF submodule)
- `unset IDF_PATH` (Clear IDF_PATH variable if it is already set)
- `./boot/espressif/hal/esp-idf/install.sh` (Installs toolchain required to build application for ESP32)
- `. ./boot/espressif/hal/esp-idf/export.sh` (Adds required directories to PATH)

## Steps to build and flash MCUboot port for ESP32-S2

- `cd boot/espressif`
- `cmake -DCMAKE_TOOLCHAIN_FILE=tools/toolchain-esp32s2.cmake -DMCUBOOT_TARGET=esp32s2 -DIDF_PATH=$IDF_PATH -B build -GNinja`
- `cmake --build build/`
- `export ESPPORT=your-esp32-serial-port`
- `$IDF_PATH/components/esptool_py/esptool/esptool.py --chip esp32s2 elf2image --flash_mode dio --flash_freq 40m -o build/mcuboot_esp32s2.bin build/mcuboot_esp32s2.elf`
- `$IDF_PATH/components/esptool_py/esptool/esptool.py -p $ESPPORT -b 2000000 --before default_reset --after hard_reset --chip esp32s2 write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 build/mcuboot_esp32s2.bin`

d3zd3z commented:
Is this ready to have the "draft" removed, now that the dependency has been merged?

almir-okato commented:
Yes @d3zd3z, please take a look. Thanks!

gustavonihei commented on docs/readme-espressif.md:
```diff
@@ -0,0 +1,70 @@
+# Building and using MCUboot with Espressif's ESP32-xx
```
```suggestion
# Building and using MCUboot with Espressif's chips.
```
I believe this is more intuitive than the `xx` placeholder.

almir-okato commented:
Last force-push was to include missing `#ifdef` regarding WDT enabling. No more changes intended.

almir-okato commented:
@d3zd3z, @utzig, could you please take a look?

d3zd3z reviewed (changes requested):
I'm fine with the changes, except that there do seem to be some stray spaces at the ends of lines. Otherwise, looks ok.

almir-okato commented:
> I'm fine with the changes, except that there do seem to be some stray spaces at the ends of lines. Otherwise, looks ok.

@d3zd3z I've removed the stray whitespaces and also updated the commit title as suggested by @gustavonihei on the other PR, please take a look.

gustavonihei reviewed (approved):
LGTM. I've already tested this port for ESP32-S2 and successfully booted a NuttX application firmware image.

utzig commented on boot/espressif/hal/CMakeLists.txt:
```diff
@@ -114,13 +124,23 @@ target_compile_options(
     -T${esp_idf_dir}/components/${MCUBOOT_TARGET}/ld/${MCUBOOT_TARGET}.peripherals.ld
     -T${esp_idf_dir}/components/esp_rom/${MCUBOOT_TARGET}/ld/${MCUBOOT_TARGET}.rom.newlib-funcs.ld
     )
+    
```
There are still spaces in this line.

almir-okato replied:
Done

utzig commented on boot/espressif/hal/src/esp32s2/bootloader_init.c:
```diff
@@ -0,0 +1,213 @@
+    uint8_t wp_gpio_num  = SPI_WP_GPIO_NUM;
+    if (spiconfig == 0) {
+
+    } else {
```
Could just test for `spiconfig != 0` and not use an `else`.

almir-okato replied:
Done

utzig commented on boot/espressif/port/esp32s2/ld/bootloader.ld:
```diff
@@ -0,0 +1,141 @@
+{
+  iram_seg (RWX) :    org = 0x40048000, len = 0x8000
+  iram_loader_seg (RWX) : org = 0x40050000, len = 0x3000
+  dram_seg (RW) :     org = 0x3FFE6000, len = 0x8000 
```
There's a trailing space here.

almir-okato replied:
Done

utzig commented on boot/espressif/port/esp32s2/ld/bootloader.ld:
```diff
@@ -0,0 +1,141 @@
+    *(.dynamic)
+    *(.gnu.version_d)
+    _rodata_end = ABSOLUTE(.);
+	/* Literals are also RO data. */
```
Change tab to spaces here.

almir-okato replied:
Done

utzig reviewed (changes requested):
I assume the "TODO"s left are to stay there, or had you planned to implement them as part of the PR?

almir-okato commented:
> I assume the "TODO"s left are to stay there, or had you planned to implement them as part of the PR?

I removed them. They were part of a code snippet from `bootloader_support` lib from the ESP-IDF.

almir-okato commented:
@utzig I've applied your comments, please take a look.

utzig commented on boot/espressif/hal/src/esp32s2/bootloader_init.c:
```diff
@@ -0,0 +1,208 @@
+    case ESP_IMAGE_FLASH_SIZE_1MB:
+        size = 1;
+        break;
+    case ESP_IMAGE_FLASH_SIZE_2MB:
```
I think this `case` did not need to here since `default` is already 2.

gustavonihei replied:
Indeed, `ESP_IMAGE_FLASH_SIZE_2MB` and the `default` case could be merged.
Most of these functions are common to both MCUboot and IDF's bootloader, which are provided by the `bootloader_support` library. But currently some of them cannot be used by MCUboot as-is.
I'd suggest for now to let this function unmodified to avoid deviating too much from the `bootloader_support` sources.

utzig reviewed (approved):
LGTM

utzig commented:
Seems like we run out of Travis credits. Since this PR does not touch `bootutil`, I'll just merge it.
